# Language Translation
In this project, you’re going to take a peek into the realm of neural network machine translation.  You’ll be training a sequence to sequence model on a dataset of English and French sentences that can translate new sentences from English to French.
## Get the Data
Since translating the whole language of English to French will take lots of time to train, we have provided you with a small portion of the English corpus.

In [1]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import helper
import problem_unittests as tests

source_path = 'data/small_vocab_en'
target_path = 'data/small_vocab_fr'
source_text = helper.load_data(source_path)
target_text = helper.load_data(target_path)

## Explore the Data
Play around with view_sentence_range to view different parts of the data.

In [2]:
view_sentence_range = (0, 10)

"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import numpy as np

print('Dataset Stats')
print('Roughly the number of unique words: {}'.format(len({word: None for word in source_text.split()})))

sentences = source_text.split('\n')
word_counts = [len(sentence.split()) for sentence in sentences]
print('Number of sentences: {}'.format(len(sentences)))
print('Average number of words in a sentence: {}'.format(np.average(word_counts)))

print()
print('English sentences {} to {}:'.format(*view_sentence_range))
print('\n'.join(source_text.split('\n')[view_sentence_range[0]:view_sentence_range[1]]))
print()
print('French sentences {} to {}:'.format(*view_sentence_range))
print('\n'.join(target_text.split('\n')[view_sentence_range[0]:view_sentence_range[1]]))

Dataset Stats
Roughly the number of unique words: 227
Number of sentences: 137861
Average number of words in a sentence: 13.225277634719028

English sentences 0 to 10:
new jersey is sometimes quiet during autumn , and it is snowy in april .
the united states is usually chilly during july , and it is usually freezing in november .
california is usually quiet during march , and it is usually hot in june .
the united states is sometimes mild during june , and it is cold in september .
your least liked fruit is the grape , but my least liked is the apple .
his favorite fruit is the orange , but my favorite is the grape .
paris is relaxing during december , but it is usually chilly in july .
new jersey is busy during spring , and it is never hot in march .
our least liked fruit is the lemon , but my least liked is the grape .
the united states is sometimes busy during january , and it is sometimes warm in november .

French sentences 0 to 10:
new jersey est parfois calme pendant l' automne 

## Implement Preprocessing Function
### Text to Word Ids
As you did with other RNNs, you must turn the text into a number so the computer can understand it. In the function `text_to_ids()`, you'll turn `source_text` and `target_text` from words to ids.  However, you need to add the `<EOS>` word id at the end of `target_text`.  This will help the neural network predict when the sentence should end.

You can get the `<EOS>` word id by doing:
```python
target_vocab_to_int['<EOS>']
```
You can get other word ids using `source_vocab_to_int` and `target_vocab_to_int`.

In [3]:
def text_to_ids(source_text, target_text, source_vocab_to_int, target_vocab_to_int):
    """
    Convert source and target text to proper word ids
    :param source_text: String that contains all the source text.
    :param target_text: String that contains all the target text.
    :param source_vocab_to_int: Dictionary to go from the source words to an id
    :param target_vocab_to_int: Dictionary to go from the target words to an id
    :return: A tuple of lists (source_id_text, target_id_text)
    """
    # TODO: Implement Function
    eos = target_vocab_to_int['<EOS>']
    source_id_text = [[source_vocab_to_int[word] for word in line.split()] for line in source_text.split('\n')]
    target_id_text = [[target_vocab_to_int[word] for word in line.split()] + [eos] for line in target_text.split('\n')]
    return (source_id_text, target_id_text)

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_text_to_ids(text_to_ids)

Tests Passed


### Preprocess all the data and save it
Running the code cell below will preprocess all the data and save it to file.

In [4]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
helper.preprocess_and_save_data(source_path, target_path, text_to_ids)

# Check Point
This is your first checkpoint. If you ever decide to come back to this notebook or have to restart the notebook, you can start from here. The preprocessed data has been saved to disk.

In [5]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import numpy as np
import helper
import problem_unittests as tests

(source_int_text, target_int_text), (source_vocab_to_int, target_vocab_to_int), _ = helper.load_preprocess()

### Check the Version of TensorFlow and Access to GPU
This will check to make sure you have the correct version of TensorFlow and access to a GPU

In [6]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
from distutils.version import LooseVersion
import warnings
import tensorflow as tf
from tensorflow.python.layers.core import Dense

# Check TensorFlow Version
assert LooseVersion(tf.__version__) >= LooseVersion('1.1'), 'Please use TensorFlow version 1.1 or newer'
print('TensorFlow Version: {}'.format(tf.__version__))

# Check for a GPU
if not tf.test.gpu_device_name():
    warnings.warn('No GPU found. Please use a GPU to train your neural network.')
else:
    print('Default GPU Device: {}'.format(tf.test.gpu_device_name()))

TensorFlow Version: 1.1.0
Default GPU Device: /gpu:0


## Build the Neural Network
You'll build the components necessary to build a Sequence-to-Sequence model by implementing the following functions below:
- `model_inputs`
- `process_decoder_input`
- `encoding_layer`
- `decoding_layer_train`
- `decoding_layer_infer`
- `decoding_layer`
- `seq2seq_model`

### Input
Implement the `model_inputs()` function to create TF Placeholders for the Neural Network. It should create the following placeholders:

- Input text placeholder named "input" using the TF Placeholder name parameter with rank 2.
- Targets placeholder with rank 2.
- Learning rate placeholder with rank 0.
- Keep probability placeholder named "keep_prob" using the TF Placeholder name parameter with rank 0.
- Target sequence length placeholder named "target_sequence_length" with rank 1
- Max target sequence length tensor named "max_target_len" getting its value from applying tf.reduce_max on the target_sequence_length placeholder. Rank 0.
- Source sequence length placeholder named "source_sequence_length" with rank 1

Return the placeholders in the following the tuple (input, targets, learning rate, keep probability, target sequence length, max target sequence length, source sequence length)

In [7]:
def model_inputs():
    """
    Create TF Placeholders for input, targets, learning rate, and lengths of source and target sequences.
    :return: Tuple (input, targets, learning rate, keep probability, target sequence length,
    max target sequence length, source sequence length)
    """
    # TODO: Implement Function
    input = tf.placeholder(tf.int32, [None,None], name='input')
    targets = tf.placeholder(tf.int32, [None,None], name='targets')
    learning_rate = tf.placeholder(tf.float32, name='learning_rate')
    keep_prob = tf.placeholder(tf.float32, name='keep_prob')
    target_sequence_length = tf.placeholder(tf.int32, [None], name='target_sequence_length')
    max_target_sequence_length = tf.reduce_max(target_sequence_length, name='max_target_sequence_length')
    source_sequence_length = tf.placeholder(tf.int32, [None], name='source_sequence_length')
    return (input, targets, learning_rate, keep_prob, target_sequence_length, max_target_sequence_length, source_sequence_length)

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_model_inputs(model_inputs)

Tests Passed


### Process Decoder Input
Implement `process_decoder_input` by removing the last word id from each batch in `target_data` and concat the GO ID to the begining of each batch.

In [8]:
def process_decoder_input(target_data, target_vocab_to_int, batch_size):
    """
    Preprocess target data for encoding
    :param target_data: Target Placehoder
    :param target_vocab_to_int: Dictionary to go from the target words to an id
    :param batch_size: Batch Size
    :return: Preprocessed target data
    """
    # TODO: Implement Function
    ending = tf.strided_slice(target_data, [0, 0], [batch_size, -1], [1, 1])
    dec_input = tf.concat([tf.fill([batch_size, 1], target_vocab_to_int['<GO>']), ending], 1)
    
    return (dec_input)

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_process_encoding_input(process_decoder_input)

Tests Passed


### Encoding
Implement `encoding_layer()` to create a Encoder RNN layer:
 * Embed the encoder input using [`tf.contrib.layers.embed_sequence`](https://www.tensorflow.org/api_docs/python/tf/contrib/layers/embed_sequence)
 * Construct a [stacked](https://github.com/tensorflow/tensorflow/blob/6947f65a374ebf29e74bb71e36fd82760056d82c/tensorflow/docs_src/tutorials/recurrent.md#stacking-multiple-lstms) [`tf.contrib.rnn.LSTMCell`](https://www.tensorflow.org/api_docs/python/tf/contrib/rnn/LSTMCell) wrapped in a [`tf.contrib.rnn.DropoutWrapper`](https://www.tensorflow.org/api_docs/python/tf/contrib/rnn/DropoutWrapper)
 * Pass cell and embedded input to [`tf.nn.dynamic_rnn()`](https://www.tensorflow.org/api_docs/python/tf/nn/dynamic_rnn)

In [18]:
from imp import reload
reload(tests)

def encoding_layer(rnn_inputs, rnn_size, num_layers, keep_prob, 
                   source_sequence_length, source_vocab_size, 
                   encoding_embedding_size):
    """
    Create encoding layer
    :param rnn_inputs: Inputs for the RNN
    :param rnn_size: RNN Size
    :param num_layers: Number of layers
    :param keep_prob: Dropout keep probability
    :param source_sequence_length: a list of the lengths of each sequence in the batch
    :param source_vocab_size: vocabulary size of source data
    :param encoding_embedding_size: embedding size of source data
    :return: tuple (RNN output, RNN state)
    """
    # TODO: Implement Function
    enc_embed_input = tf.contrib.layers.embed_sequence(rnn_inputs, source_vocab_size, encoding_embedding_size)
    #cell = tf.contrib.rnn.LSTMCell(rnn_size)
    #cell = tf.contrib.rnn.DropoutWrapper(cell, output_keep_prob=keep_prob)
    enc_cell = tf.contrib.rnn.MultiRNNCell([tf.contrib.rnn.DropoutWrapper(tf.contrib.rnn.LSTMCell(rnn_size), output_keep_prob=keep_prob) for _ in range(num_layers)])
    RNN_output, RNN_state = tf.nn.dynamic_rnn(enc_cell, enc_embed_input, source_sequence_length, dtype=tf.float32)
    return (RNN_output, RNN_state)

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_encoding_layer(encoding_layer)

Tests Passed


### Decoding - Training
Create a training decoding layer:
* Create a [`tf.contrib.seq2seq.TrainingHelper`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/TrainingHelper) 
* Create a [`tf.contrib.seq2seq.BasicDecoder`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/BasicDecoder)
* Obtain the decoder outputs from [`tf.contrib.seq2seq.dynamic_decode`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/dynamic_decode)

In [20]:

def decoding_layer_train(encoder_state, dec_cell, dec_embed_input, 
                         target_sequence_length, max_summary_length, 
                         output_layer, keep_prob):
    """
    Create a decoding layer for training
    :param encoder_state: Encoder State
    :param dec_cell: Decoder RNN Cell
    :param dec_embed_input: Decoder embedded input
    :param target_sequence_length: The lengths of each sequence in the target batch
    :param max_summary_length: The length of the longest sequence in the batch
    :param output_layer: Function to apply the output layer
    :param keep_prob: Dropout keep probability
    :return: BasicDecoderOutput containing training logits and sample_id
    """
    # TODO: Implement Function
    training_helper = tf.contrib.seq2seq.TrainingHelper(inputs=dec_embed_input, sequence_length=target_sequence_length, time_major=False)
    training_decoder = tf.contrib.seq2seq.BasicDecoder(dec_cell, training_helper, encoder_state, output_layer)
    training_decoder_output = tf.contrib.seq2seq.dynamic_decode(training_decoder,
                                                                       impute_finished=True,
                                                                       maximum_iterations=max_summary_length)[0]
    return(training_decoder_output)



"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_decoding_layer_train(decoding_layer_train)

Tests Passed


### Decoding - Inference
Create inference decoder:
* Create a [`tf.contrib.seq2seq.GreedyEmbeddingHelper`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/GreedyEmbeddingHelper)
* Create a [`tf.contrib.seq2seq.BasicDecoder`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/BasicDecoder)
* Obtain the decoder outputs from [`tf.contrib.seq2seq.dynamic_decode`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/dynamic_decode)

In [24]:
def decoding_layer_infer(encoder_state, dec_cell, dec_embeddings, start_of_sequence_id,
                         end_of_sequence_id, max_target_sequence_length,
                         vocab_size, output_layer, batch_size, keep_prob):
    """
    Create a decoding layer for inference
    :param encoder_state: Encoder state
    :param dec_cell: Decoder RNN Cell
    :param dec_embeddings: Decoder embeddings
    :param start_of_sequence_id: GO ID
    :param end_of_sequence_id: EOS Id
    :param max_target_sequence_length: Maximum length of target sequences
    :param vocab_size: Size of decoder/target vocabulary
    :param decoding_scope: TenorFlow Variable Scope for decoding
    :param output_layer: Function to apply the output layer
    :param batch_size: Batch size
    :param keep_prob: Dropout keep probability
    :return: BasicDecoderOutput containing inference logits and sample_id
    """
    # TODO: Implement Function
    start_tokens = tf.tile(tf.constant([start_of_sequence_id], dtype=tf.int32), [batch_size], name='start_tokens')
    inference_helper = tf.contrib.seq2seq.GreedyEmbeddingHelper(dec_embeddings,
                                                                start_tokens,
                                                                end_of_sequence_id)
    inference_decoder = tf.contrib.seq2seq.BasicDecoder(dec_cell,
                                                        inference_helper,
                                                        encoder_state,
                                                        output_layer)
    inference_decoder_output = tf.contrib.seq2seq.dynamic_decode(inference_decoder,
                                                            impute_finished=True,
                                                            maximum_iterations=max_target_sequence_length)[0]
    return (inference_decoder_output)



"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_decoding_layer_infer(decoding_layer_infer)

Tests Passed


### Build the Decoding Layer
Implement `decoding_layer()` to create a Decoder RNN layer.

* Embed the target sequences
* Construct the decoder LSTM cell (just like you constructed the encoder cell above)
* Create an output layer to map the outputs of the decoder to the elements of our vocabulary
* Use the your `decoding_layer_train(encoder_state, dec_cell, dec_embed_input, target_sequence_length, max_target_sequence_length, output_layer, keep_prob)` function to get the training logits.
* Use your `decoding_layer_infer(encoder_state, dec_cell, dec_embeddings, start_of_sequence_id, end_of_sequence_id, max_target_sequence_length, vocab_size, output_layer, batch_size, keep_prob)` function to get the inference logits.

Note: You'll need to use [tf.variable_scope](https://www.tensorflow.org/api_docs/python/tf/variable_scope) to share variables between training and inference.

In [35]:
def decoding_layer(dec_input, encoder_state,
                   target_sequence_length, max_target_sequence_length,
                   rnn_size,
                   num_layers, target_vocab_to_int, target_vocab_size,
                   batch_size, keep_prob, decoding_embedding_size):
    """
    Create decoding layer
    :param dec_input: Decoder input
    :param encoder_state: Encoder state
    :param target_sequence_length: The lengths of each sequence in the target batch
    :param max_target_sequence_length: Maximum length of target sequences
    :param rnn_size: RNN Size
    :param num_layers: Number of layers
    :param target_vocab_to_int: Dictionary to go from the target words to an id
    :param target_vocab_size: Size of target vocabulary
    :param batch_size: The size of the batch
    :param keep_prob: Dropout keep probability
    :param decoding_embedding_size: Decoding embedding size
    :return: Tuple of (Training BasicDecoderOutput, Inference BasicDecoderOutput)
    """
    # TODO: Implement Function
    start_of_sequence_id = target_vocab_to_int['<GO>']
    end_of_sequence_id = target_vocab_to_int['<EOS>']
    
    dec_embeddings = tf.Variable(tf.random_uniform([target_vocab_size, decoding_embedding_size]))
    dec_embed_input = tf.nn.embedding_lookup(dec_embeddings, dec_input)
    
    cell = tf.contrib.rnn.MultiRNNCell([tf.contrib.rnn.DropoutWrapper(tf.contrib.rnn.LSTMCell(rnn_size), output_keep_prob=keep_prob) for _ in range(num_layers)])
    
    #with tf.variable_scope("decoding") as decoding_scope:
    #    output_fn = lambda x: tf.contrib.layers.fully_connected(x, target_vocab_size, None, scope=decoding_scope)
    output_fn = Dense(target_vocab_size, kernel_initializer = tf.truncated_normal_initializer(mean = 0.0, stddev=0.1))
 

    with tf.variable_scope('decoding'):
        train_logits = decoding_layer_train(encoder_state, cell, dec_embed_input, target_sequence_length, max_target_sequence_length, 
                                      output_fn, keep_prob)   
        
    with tf.variable_scope('decoding', reuse=True) as decoding_scope:
        infer_logits = decoding_layer_infer(encoder_state, cell, dec_embeddings, start_of_sequence_id,
                                        end_of_sequence_id, max_target_sequence_length, target_vocab_size,
                                        output_fn, batch_size, keep_prob)
    return train_logits, infer_logits

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_decoding_layer(decoding_layer)

Tests Passed


### Build the Neural Network
Apply the functions you implemented above to:

- Encode the input using your `encoding_layer(rnn_inputs, rnn_size, num_layers, keep_prob,  source_sequence_length, source_vocab_size, encoding_embedding_size)`.
- Process target data using your `process_decoder_input(target_data, target_vocab_to_int, batch_size)` function.
- Decode the encoded input using your `decoding_layer(dec_input, enc_state, target_sequence_length, max_target_sentence_length, rnn_size, num_layers, target_vocab_to_int, target_vocab_size, batch_size, keep_prob, dec_embedding_size)` function.

In [38]:
def seq2seq_model(input_data, target_data, keep_prob, batch_size,
                  source_sequence_length, target_sequence_length,
                  max_target_sentence_length,
                  source_vocab_size, target_vocab_size,
                  enc_embedding_size, dec_embedding_size,
                  rnn_size, num_layers, target_vocab_to_int):
    """
    Build the Sequence-to-Sequence part of the neural network
    :param input_data: Input placeholder
    :param target_data: Target placeholder
    :param keep_prob: Dropout keep probability placeholder
    :param batch_size: Batch Size
    :param source_sequence_length: Sequence Lengths of source sequences in the batch
    :param target_sequence_length: Sequence Lengths of target sequences in the batch
    :param source_vocab_size: Source vocabulary size
    :param target_vocab_size: Target vocabulary size
    :param enc_embedding_size: Decoder embedding size
    :param dec_embedding_size: Encoder embedding size
    :param rnn_size: RNN Size
    :param num_layers: Number of layers
    :param target_vocab_to_int: Dictionary to go from the target words to an id
    :return: Tuple of (Training BasicDecoderOutput, Inference BasicDecoderOutput)
    """
    # TODO: Implement Function
    
    _, enc_state = encoding_layer(input_data, 
                                  rnn_size, 
                                  num_layers, 
                                  keep_prob,
                                  source_sequence_length,
                                  source_vocab_size, 
                                  enc_embedding_size)
    
    
    dec_input = process_decoder_input(target_data, target_vocab_to_int, batch_size)

    training_decoder_output, inference_decoder_output = decoding_layer(dec_input, enc_state,
                                                                   target_sequence_length, max_target_sentence_length,
                                                                   rnn_size,
                                                                   num_layers, target_vocab_to_int, target_vocab_size,
                                                                   batch_size, keep_prob, dec_embedding_size)
    return (training_decoder_output, inference_decoder_output)


"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_seq2seq_model(seq2seq_model)

Tests Passed


## Neural Network Training
### Hyperparameters
Tune the following parameters:

- Set `epochs` to the number of epochs.
- Set `batch_size` to the batch size.
- Set `rnn_size` to the size of the RNNs.
- Set `num_layers` to the number of layers.
- Set `encoding_embedding_size` to the size of the embedding for the encoder.
- Set `decoding_embedding_size` to the size of the embedding for the decoder.
- Set `learning_rate` to the learning rate.
- Set `keep_probability` to the Dropout keep probability
- Set `display_step` to state how many steps between each debug output statement

In [52]:
# Number of Epochs
epochs = 10
# Batch Size
batch_size = 512
# RNN Size
rnn_size = 256
# Number of Layers
num_layers = 2
# Embedding Size
encoding_embedding_size = 200
decoding_embedding_size = 200
# Learning Rate
learning_rate = 0.01
# Dropout Keep Probability
keep_probability = 0.5
display_step = 2

### Build the Graph
Build the graph using the neural network you implemented.

In [49]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
save_path = 'checkpoints/dev'
(source_int_text, target_int_text), (source_vocab_to_int, target_vocab_to_int), _ = helper.load_preprocess()
max_target_sentence_length = max([len(sentence) for sentence in source_int_text])

train_graph = tf.Graph()
with train_graph.as_default():
    input_data, targets, lr, keep_prob, target_sequence_length, max_target_sequence_length, source_sequence_length = model_inputs()

    #sequence_length = tf.placeholder_with_default(max_target_sentence_length, None, name='sequence_length')
    input_shape = tf.shape(input_data)

    train_logits, inference_logits = seq2seq_model(tf.reverse(input_data, [-1]),
                                                   targets,
                                                   keep_prob,
                                                   batch_size,
                                                   source_sequence_length,
                                                   target_sequence_length,
                                                   max_target_sequence_length,
                                                   len(source_vocab_to_int),
                                                   len(target_vocab_to_int),
                                                   encoding_embedding_size,
                                                   decoding_embedding_size,
                                                   rnn_size,
                                                   num_layers,
                                                   target_vocab_to_int)


    training_logits = tf.identity(train_logits.rnn_output, name='logits')
    inference_logits = tf.identity(inference_logits.sample_id, name='predictions')

    masks = tf.sequence_mask(target_sequence_length, max_target_sequence_length, dtype=tf.float32, name='masks')

    with tf.name_scope("optimization"):
        # Loss function
        cost = tf.contrib.seq2seq.sequence_loss(
            training_logits,
            targets,
            masks)

        # Optimizer
        optimizer = tf.train.AdamOptimizer(lr)

        # Gradient Clipping
        gradients = optimizer.compute_gradients(cost)
        capped_gradients = [(tf.clip_by_value(grad, -1., 1.), var) for grad, var in gradients if grad is not None]
        train_op = optimizer.apply_gradients(capped_gradients)


Batch and pad the source and target sequences

In [50]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
def pad_sentence_batch(sentence_batch, pad_int):
    """Pad sentences with <PAD> so that each sentence of a batch has the same length"""
    max_sentence = max([len(sentence) for sentence in sentence_batch])
    return [sentence + [pad_int] * (max_sentence - len(sentence)) for sentence in sentence_batch]


def get_batches(sources, targets, batch_size, source_pad_int, target_pad_int):
    """Batch targets, sources, and the lengths of their sentences together"""
    for batch_i in range(0, len(sources)//batch_size):
        start_i = batch_i * batch_size

        # Slice the right amount for the batch
        sources_batch = sources[start_i:start_i + batch_size]
        targets_batch = targets[start_i:start_i + batch_size]

        # Pad
        pad_sources_batch = np.array(pad_sentence_batch(sources_batch, source_pad_int))
        pad_targets_batch = np.array(pad_sentence_batch(targets_batch, target_pad_int))

        # Need the lengths for the _lengths parameters
        pad_targets_lengths = []
        for target in pad_targets_batch:
            pad_targets_lengths.append(len(target))

        pad_source_lengths = []
        for source in pad_sources_batch:
            pad_source_lengths.append(len(source))

        yield pad_sources_batch, pad_targets_batch, pad_source_lengths, pad_targets_lengths


### Train
Train the neural network on the preprocessed data. If you have a hard time getting a good loss, check the forms to see if anyone is having the same problem.

In [53]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
def get_accuracy(target, logits):
    """
    Calculate accuracy
    """
    max_seq = max(target.shape[1], logits.shape[1])
    if max_seq - target.shape[1]:
        target = np.pad(
            target,
            [(0,0),(0,max_seq - target.shape[1])],
            'constant')
    if max_seq - logits.shape[1]:
        logits = np.pad(
            logits,
            [(0,0),(0,max_seq - logits.shape[1])],
            'constant')

    return np.mean(np.equal(target, logits))

# Split data to training and validation sets
train_source = source_int_text[batch_size:]
train_target = target_int_text[batch_size:]
valid_source = source_int_text[:batch_size]
valid_target = target_int_text[:batch_size]
(valid_sources_batch, valid_targets_batch, valid_sources_lengths, valid_targets_lengths ) = next(get_batches(valid_source,
                                                                                                             valid_target,
                                                                                                             batch_size,
                                                                                                             source_vocab_to_int['<PAD>'],
                                                                                                             target_vocab_to_int['<PAD>']))                                                                                                  
with tf.Session(graph=train_graph) as sess:
    sess.run(tf.global_variables_initializer())

    for epoch_i in range(epochs):
        for batch_i, (source_batch, target_batch, sources_lengths, targets_lengths) in enumerate(
                get_batches(train_source, train_target, batch_size,
                            source_vocab_to_int['<PAD>'],
                            target_vocab_to_int['<PAD>'])):

            _, loss = sess.run(
                [train_op, cost],
                {input_data: source_batch,
                 targets: target_batch,
                 lr: learning_rate,
                 target_sequence_length: targets_lengths,
                 source_sequence_length: sources_lengths,
                 keep_prob: keep_probability})


            if batch_i % display_step == 0 and batch_i > 0:


                batch_train_logits = sess.run(
                    inference_logits,
                    {input_data: source_batch,
                     source_sequence_length: sources_lengths,
                     target_sequence_length: targets_lengths,
                     keep_prob: 1.0})


                batch_valid_logits = sess.run(
                    inference_logits,
                    {input_data: valid_sources_batch,
                     source_sequence_length: valid_sources_lengths,
                     target_sequence_length: valid_targets_lengths,
                     keep_prob: 1.0})

                train_acc = get_accuracy(target_batch, batch_train_logits)

                valid_acc = get_accuracy(valid_targets_batch, batch_valid_logits)

                print('Epoch {:>3} Batch {:>4}/{} - Train Accuracy: {:>6.4f}, Validation Accuracy: {:>6.4f}, Loss: {:>6.4f}'
                      .format(epoch_i, batch_i, len(source_int_text) // batch_size, train_acc, valid_acc, loss))

    # Save Model
    saver = tf.train.Saver()
    saver.save(sess, save_path)
    print('Model Trained and Saved')

Epoch   0 Batch    2/269 - Train Accuracy: 0.0701, Validation Accuracy: 0.0645, Loss: 4.6229
Epoch   0 Batch    4/269 - Train Accuracy: 0.2681, Validation Accuracy: 0.3409, Loss: 4.0785
Epoch   0 Batch    6/269 - Train Accuracy: 0.3129, Validation Accuracy: 0.3420, Loss: 3.7748
Epoch   0 Batch    8/269 - Train Accuracy: 0.2831, Validation Accuracy: 0.3461, Loss: 3.6634
Epoch   0 Batch   10/269 - Train Accuracy: 0.2803, Validation Accuracy: 0.3501, Loss: 3.5200
Epoch   0 Batch   12/269 - Train Accuracy: 0.2921, Validation Accuracy: 0.3518, Loss: 3.3963
Epoch   0 Batch   14/269 - Train Accuracy: 0.3472, Validation Accuracy: 0.3818, Loss: 3.1455
Epoch   0 Batch   16/269 - Train Accuracy: 0.3603, Validation Accuracy: 0.3886, Loss: 3.0185
Epoch   0 Batch   18/269 - Train Accuracy: 0.3444, Validation Accuracy: 0.4050, Loss: 3.0250
Epoch   0 Batch   20/269 - Train Accuracy: 0.3535, Validation Accuracy: 0.4129, Loss: 2.9217
Epoch   0 Batch   22/269 - Train Accuracy: 0.3956, Validation Accuracy

Epoch   0 Batch  180/269 - Train Accuracy: 0.6505, Validation Accuracy: 0.6667, Loss: 0.5708
Epoch   0 Batch  182/269 - Train Accuracy: 0.6576, Validation Accuracy: 0.6656, Loss: 0.5779
Epoch   0 Batch  184/269 - Train Accuracy: 0.6466, Validation Accuracy: 0.6659, Loss: 0.5984
Epoch   0 Batch  186/269 - Train Accuracy: 0.6379, Validation Accuracy: 0.6641, Loss: 0.5866
Epoch   0 Batch  188/269 - Train Accuracy: 0.6570, Validation Accuracy: 0.6554, Loss: 0.5544
Epoch   0 Batch  190/269 - Train Accuracy: 0.6664, Validation Accuracy: 0.6713, Loss: 0.5514
Epoch   0 Batch  192/269 - Train Accuracy: 0.6707, Validation Accuracy: 0.6738, Loss: 0.5643
Epoch   0 Batch  194/269 - Train Accuracy: 0.6637, Validation Accuracy: 0.6667, Loss: 0.5549
Epoch   0 Batch  196/269 - Train Accuracy: 0.6286, Validation Accuracy: 0.6585, Loss: 0.5505
Epoch   0 Batch  198/269 - Train Accuracy: 0.6479, Validation Accuracy: 0.6665, Loss: 0.5733
Epoch   0 Batch  200/269 - Train Accuracy: 0.6679, Validation Accuracy

Epoch   1 Batch   92/269 - Train Accuracy: 0.7853, Validation Accuracy: 0.7828, Loss: 0.3127
Epoch   1 Batch   94/269 - Train Accuracy: 0.7658, Validation Accuracy: 0.7896, Loss: 0.3277
Epoch   1 Batch   96/269 - Train Accuracy: 0.7695, Validation Accuracy: 0.7860, Loss: 0.3052
Epoch   1 Batch   98/269 - Train Accuracy: 0.7763, Validation Accuracy: 0.7904, Loss: 0.3086
Epoch   1 Batch  100/269 - Train Accuracy: 0.8166, Validation Accuracy: 0.7954, Loss: 0.3003
Epoch   1 Batch  102/269 - Train Accuracy: 0.7822, Validation Accuracy: 0.7909, Loss: 0.2928
Epoch   1 Batch  104/269 - Train Accuracy: 0.7968, Validation Accuracy: 0.7920, Loss: 0.2890
Epoch   1 Batch  106/269 - Train Accuracy: 0.7826, Validation Accuracy: 0.7879, Loss: 0.2826
Epoch   1 Batch  108/269 - Train Accuracy: 0.7889, Validation Accuracy: 0.7927, Loss: 0.2890
Epoch   1 Batch  110/269 - Train Accuracy: 0.7892, Validation Accuracy: 0.7920, Loss: 0.2793
Epoch   1 Batch  112/269 - Train Accuracy: 0.7957, Validation Accuracy

Epoch   2 Batch    4/269 - Train Accuracy: 0.8924, Validation Accuracy: 0.8960, Loss: 0.1427
Epoch   2 Batch    6/269 - Train Accuracy: 0.9130, Validation Accuracy: 0.9071, Loss: 0.1362
Epoch   2 Batch    8/269 - Train Accuracy: 0.9062, Validation Accuracy: 0.9060, Loss: 0.1542
Epoch   2 Batch   10/269 - Train Accuracy: 0.8994, Validation Accuracy: 0.9066, Loss: 0.1359
Epoch   2 Batch   12/269 - Train Accuracy: 0.8875, Validation Accuracy: 0.9037, Loss: 0.1466
Epoch   2 Batch   14/269 - Train Accuracy: 0.8902, Validation Accuracy: 0.9070, Loss: 0.1362
Epoch   2 Batch   16/269 - Train Accuracy: 0.8908, Validation Accuracy: 0.9037, Loss: 0.1447
Epoch   2 Batch   18/269 - Train Accuracy: 0.8990, Validation Accuracy: 0.9030, Loss: 0.1422
Epoch   2 Batch   20/269 - Train Accuracy: 0.9162, Validation Accuracy: 0.9074, Loss: 0.1326
Epoch   2 Batch   22/269 - Train Accuracy: 0.9117, Validation Accuracy: 0.9110, Loss: 0.1264
Epoch   2 Batch   24/269 - Train Accuracy: 0.8984, Validation Accuracy

Epoch   2 Batch  182/269 - Train Accuracy: 0.9246, Validation Accuracy: 0.9312, Loss: 0.0839
Epoch   2 Batch  184/269 - Train Accuracy: 0.9248, Validation Accuracy: 0.9375, Loss: 0.0780
Epoch   2 Batch  186/269 - Train Accuracy: 0.9267, Validation Accuracy: 0.9345, Loss: 0.0802
Epoch   2 Batch  188/269 - Train Accuracy: 0.9312, Validation Accuracy: 0.9391, Loss: 0.0711
Epoch   2 Batch  190/269 - Train Accuracy: 0.9237, Validation Accuracy: 0.9423, Loss: 0.0758
Epoch   2 Batch  192/269 - Train Accuracy: 0.9289, Validation Accuracy: 0.9379, Loss: 0.0809
Epoch   2 Batch  194/269 - Train Accuracy: 0.9213, Validation Accuracy: 0.9384, Loss: 0.0748
Epoch   2 Batch  196/269 - Train Accuracy: 0.9308, Validation Accuracy: 0.9407, Loss: 0.0729
Epoch   2 Batch  198/269 - Train Accuracy: 0.9204, Validation Accuracy: 0.9461, Loss: 0.0787
Epoch   2 Batch  200/269 - Train Accuracy: 0.9272, Validation Accuracy: 0.9403, Loss: 0.0709
Epoch   2 Batch  202/269 - Train Accuracy: 0.9233, Validation Accuracy

Epoch   3 Batch   94/269 - Train Accuracy: 0.9389, Validation Accuracy: 0.9476, Loss: 0.0788
Epoch   3 Batch   96/269 - Train Accuracy: 0.9306, Validation Accuracy: 0.9478, Loss: 0.0648
Epoch   3 Batch   98/269 - Train Accuracy: 0.9465, Validation Accuracy: 0.9484, Loss: 0.0596
Epoch   3 Batch  100/269 - Train Accuracy: 0.9462, Validation Accuracy: 0.9495, Loss: 0.0611
Epoch   3 Batch  102/269 - Train Accuracy: 0.9439, Validation Accuracy: 0.9552, Loss: 0.0549
Epoch   3 Batch  104/269 - Train Accuracy: 0.9578, Validation Accuracy: 0.9496, Loss: 0.0563
Epoch   3 Batch  106/269 - Train Accuracy: 0.9505, Validation Accuracy: 0.9433, Loss: 0.0510
Epoch   3 Batch  108/269 - Train Accuracy: 0.9442, Validation Accuracy: 0.9387, Loss: 0.0580
Epoch   3 Batch  110/269 - Train Accuracy: 0.9323, Validation Accuracy: 0.9369, Loss: 0.0577
Epoch   3 Batch  112/269 - Train Accuracy: 0.9413, Validation Accuracy: 0.9437, Loss: 0.0636
Epoch   3 Batch  114/269 - Train Accuracy: 0.9397, Validation Accuracy

Epoch   4 Batch    6/269 - Train Accuracy: 0.9516, Validation Accuracy: 0.9535, Loss: 0.0434
Epoch   4 Batch    8/269 - Train Accuracy: 0.9599, Validation Accuracy: 0.9561, Loss: 0.0518
Epoch   4 Batch   10/269 - Train Accuracy: 0.9454, Validation Accuracy: 0.9541, Loss: 0.0470
Epoch   4 Batch   12/269 - Train Accuracy: 0.9318, Validation Accuracy: 0.9509, Loss: 0.0577
Epoch   4 Batch   14/269 - Train Accuracy: 0.9408, Validation Accuracy: 0.9470, Loss: 0.0525
Epoch   4 Batch   16/269 - Train Accuracy: 0.9343, Validation Accuracy: 0.9507, Loss: 0.0553
Epoch   4 Batch   18/269 - Train Accuracy: 0.9384, Validation Accuracy: 0.9568, Loss: 0.0536
Epoch   4 Batch   20/269 - Train Accuracy: 0.9532, Validation Accuracy: 0.9589, Loss: 0.0492
Epoch   4 Batch   22/269 - Train Accuracy: 0.9558, Validation Accuracy: 0.9566, Loss: 0.0472
Epoch   4 Batch   24/269 - Train Accuracy: 0.9416, Validation Accuracy: 0.9506, Loss: 0.0501
Epoch   4 Batch   26/269 - Train Accuracy: 0.9488, Validation Accuracy

Epoch   4 Batch  184/269 - Train Accuracy: 0.9583, Validation Accuracy: 0.9545, Loss: 0.0415
Epoch   4 Batch  186/269 - Train Accuracy: 0.9531, Validation Accuracy: 0.9545, Loss: 0.0402
Epoch   4 Batch  188/269 - Train Accuracy: 0.9569, Validation Accuracy: 0.9572, Loss: 0.0436
Epoch   4 Batch  190/269 - Train Accuracy: 0.9381, Validation Accuracy: 0.9581, Loss: 0.0485
Epoch   4 Batch  192/269 - Train Accuracy: 0.9596, Validation Accuracy: 0.9543, Loss: 0.0445
Epoch   4 Batch  194/269 - Train Accuracy: 0.9466, Validation Accuracy: 0.9545, Loss: 0.0428
Epoch   4 Batch  196/269 - Train Accuracy: 0.9404, Validation Accuracy: 0.9582, Loss: 0.0419
Epoch   4 Batch  198/269 - Train Accuracy: 0.9558, Validation Accuracy: 0.9543, Loss: 0.0492
Epoch   4 Batch  200/269 - Train Accuracy: 0.9597, Validation Accuracy: 0.9560, Loss: 0.0415
Epoch   4 Batch  202/269 - Train Accuracy: 0.9536, Validation Accuracy: 0.9554, Loss: 0.0407
Epoch   4 Batch  204/269 - Train Accuracy: 0.9599, Validation Accuracy

Epoch   5 Batch   96/269 - Train Accuracy: 0.9465, Validation Accuracy: 0.9568, Loss: 0.0458
Epoch   5 Batch   98/269 - Train Accuracy: 0.9510, Validation Accuracy: 0.9512, Loss: 0.0404
Epoch   5 Batch  100/269 - Train Accuracy: 0.9543, Validation Accuracy: 0.9604, Loss: 0.0439
Epoch   5 Batch  102/269 - Train Accuracy: 0.9572, Validation Accuracy: 0.9611, Loss: 0.0427
Epoch   5 Batch  104/269 - Train Accuracy: 0.9664, Validation Accuracy: 0.9569, Loss: 0.0428
Epoch   5 Batch  106/269 - Train Accuracy: 0.9628, Validation Accuracy: 0.9604, Loss: 0.0363
Epoch   5 Batch  108/269 - Train Accuracy: 0.9640, Validation Accuracy: 0.9608, Loss: 0.0430
Epoch   5 Batch  110/269 - Train Accuracy: 0.9544, Validation Accuracy: 0.9588, Loss: 0.0376
Epoch   5 Batch  112/269 - Train Accuracy: 0.9533, Validation Accuracy: 0.9594, Loss: 0.0412
Epoch   5 Batch  114/269 - Train Accuracy: 0.9488, Validation Accuracy: 0.9602, Loss: 0.0397
Epoch   5 Batch  116/269 - Train Accuracy: 0.9673, Validation Accuracy

Epoch   6 Batch    8/269 - Train Accuracy: 0.9640, Validation Accuracy: 0.9628, Loss: 0.0433
Epoch   6 Batch   10/269 - Train Accuracy: 0.9628, Validation Accuracy: 0.9642, Loss: 0.0342
Epoch   6 Batch   12/269 - Train Accuracy: 0.9418, Validation Accuracy: 0.9651, Loss: 0.0460
Epoch   6 Batch   14/269 - Train Accuracy: 0.9611, Validation Accuracy: 0.9600, Loss: 0.0399
Epoch   6 Batch   16/269 - Train Accuracy: 0.9502, Validation Accuracy: 0.9563, Loss: 0.0455
Epoch   6 Batch   18/269 - Train Accuracy: 0.9505, Validation Accuracy: 0.9588, Loss: 0.0470
Epoch   6 Batch   20/269 - Train Accuracy: 0.9598, Validation Accuracy: 0.9606, Loss: 0.0395
Epoch   6 Batch   22/269 - Train Accuracy: 0.9656, Validation Accuracy: 0.9575, Loss: 0.0372
Epoch   6 Batch   24/269 - Train Accuracy: 0.9560, Validation Accuracy: 0.9592, Loss: 0.0378
Epoch   6 Batch   26/269 - Train Accuracy: 0.9541, Validation Accuracy: 0.9547, Loss: 0.0351
Epoch   6 Batch   28/269 - Train Accuracy: 0.9405, Validation Accuracy

Epoch   6 Batch  186/269 - Train Accuracy: 0.9665, Validation Accuracy: 0.9553, Loss: 0.0372
Epoch   6 Batch  188/269 - Train Accuracy: 0.9535, Validation Accuracy: 0.9575, Loss: 0.0364
Epoch   6 Batch  190/269 - Train Accuracy: 0.9554, Validation Accuracy: 0.9598, Loss: 0.0409
Epoch   6 Batch  192/269 - Train Accuracy: 0.9676, Validation Accuracy: 0.9578, Loss: 0.0405
Epoch   6 Batch  194/269 - Train Accuracy: 0.9541, Validation Accuracy: 0.9600, Loss: 0.0415
Epoch   6 Batch  196/269 - Train Accuracy: 0.9504, Validation Accuracy: 0.9554, Loss: 0.0441
Epoch   6 Batch  198/269 - Train Accuracy: 0.9650, Validation Accuracy: 0.9602, Loss: 0.0401
Epoch   6 Batch  200/269 - Train Accuracy: 0.9653, Validation Accuracy: 0.9608, Loss: 0.0389
Epoch   6 Batch  202/269 - Train Accuracy: 0.9596, Validation Accuracy: 0.9607, Loss: 0.0403
Epoch   6 Batch  204/269 - Train Accuracy: 0.9558, Validation Accuracy: 0.9588, Loss: 0.0381
Epoch   6 Batch  206/269 - Train Accuracy: 0.9402, Validation Accuracy

Epoch   7 Batch   98/269 - Train Accuracy: 0.9648, Validation Accuracy: 0.9576, Loss: 0.0393
Epoch   7 Batch  100/269 - Train Accuracy: 0.9669, Validation Accuracy: 0.9577, Loss: 0.0374
Epoch   7 Batch  102/269 - Train Accuracy: 0.9598, Validation Accuracy: 0.9596, Loss: 0.0396
Epoch   7 Batch  104/269 - Train Accuracy: 0.9653, Validation Accuracy: 0.9567, Loss: 0.0362
Epoch   7 Batch  106/269 - Train Accuracy: 0.9722, Validation Accuracy: 0.9512, Loss: 0.0307
Epoch   7 Batch  108/269 - Train Accuracy: 0.9724, Validation Accuracy: 0.9516, Loss: 0.0298
Epoch   7 Batch  110/269 - Train Accuracy: 0.9593, Validation Accuracy: 0.9521, Loss: 0.0291
Epoch   7 Batch  112/269 - Train Accuracy: 0.9574, Validation Accuracy: 0.9541, Loss: 0.0363
Epoch   7 Batch  114/269 - Train Accuracy: 0.9580, Validation Accuracy: 0.9543, Loss: 0.0383
Epoch   7 Batch  116/269 - Train Accuracy: 0.9686, Validation Accuracy: 0.9585, Loss: 0.0343
Epoch   7 Batch  118/269 - Train Accuracy: 0.9652, Validation Accuracy

Epoch   8 Batch   10/269 - Train Accuracy: 0.9685, Validation Accuracy: 0.9616, Loss: 0.0308
Epoch   8 Batch   12/269 - Train Accuracy: 0.9549, Validation Accuracy: 0.9618, Loss: 0.0391
Epoch   8 Batch   14/269 - Train Accuracy: 0.9620, Validation Accuracy: 0.9624, Loss: 0.0340
Epoch   8 Batch   16/269 - Train Accuracy: 0.9538, Validation Accuracy: 0.9629, Loss: 0.0410
Epoch   8 Batch   18/269 - Train Accuracy: 0.9594, Validation Accuracy: 0.9621, Loss: 0.0308
Epoch   8 Batch   20/269 - Train Accuracy: 0.9681, Validation Accuracy: 0.9648, Loss: 0.0336
Epoch   8 Batch   22/269 - Train Accuracy: 0.9634, Validation Accuracy: 0.9605, Loss: 0.0317
Epoch   8 Batch   24/269 - Train Accuracy: 0.9631, Validation Accuracy: 0.9578, Loss: 0.0333
Epoch   8 Batch   26/269 - Train Accuracy: 0.9584, Validation Accuracy: 0.9594, Loss: 0.0327
Epoch   8 Batch   28/269 - Train Accuracy: 0.9511, Validation Accuracy: 0.9605, Loss: 0.0316
Epoch   8 Batch   30/269 - Train Accuracy: 0.9627, Validation Accuracy

Epoch   8 Batch  188/269 - Train Accuracy: 0.9634, Validation Accuracy: 0.9674, Loss: 0.0308
Epoch   8 Batch  190/269 - Train Accuracy: 0.9661, Validation Accuracy: 0.9693, Loss: 0.0330
Epoch   8 Batch  192/269 - Train Accuracy: 0.9712, Validation Accuracy: 0.9688, Loss: 0.0308
Epoch   8 Batch  194/269 - Train Accuracy: 0.9605, Validation Accuracy: 0.9680, Loss: 0.0324
Epoch   8 Batch  196/269 - Train Accuracy: 0.9555, Validation Accuracy: 0.9642, Loss: 0.0337
Epoch   8 Batch  198/269 - Train Accuracy: 0.9698, Validation Accuracy: 0.9646, Loss: 0.0324
Epoch   8 Batch  200/269 - Train Accuracy: 0.9707, Validation Accuracy: 0.9667, Loss: 0.0309
Epoch   8 Batch  202/269 - Train Accuracy: 0.9593, Validation Accuracy: 0.9684, Loss: 0.0364
Epoch   8 Batch  204/269 - Train Accuracy: 0.9740, Validation Accuracy: 0.9681, Loss: 0.0306
Epoch   8 Batch  206/269 - Train Accuracy: 0.9547, Validation Accuracy: 0.9652, Loss: 0.0397
Epoch   8 Batch  208/269 - Train Accuracy: 0.9670, Validation Accuracy

Epoch   9 Batch  100/269 - Train Accuracy: 0.9611, Validation Accuracy: 0.9720, Loss: 0.0319
Epoch   9 Batch  102/269 - Train Accuracy: 0.9705, Validation Accuracy: 0.9685, Loss: 0.0321
Epoch   9 Batch  104/269 - Train Accuracy: 0.9687, Validation Accuracy: 0.9674, Loss: 0.0360
Epoch   9 Batch  106/269 - Train Accuracy: 0.9674, Validation Accuracy: 0.9637, Loss: 0.0278
Epoch   9 Batch  108/269 - Train Accuracy: 0.9701, Validation Accuracy: 0.9664, Loss: 0.0345
Epoch   9 Batch  110/269 - Train Accuracy: 0.9659, Validation Accuracy: 0.9703, Loss: 0.0299
Epoch   9 Batch  112/269 - Train Accuracy: 0.9710, Validation Accuracy: 0.9694, Loss: 0.0321
Epoch   9 Batch  114/269 - Train Accuracy: 0.9619, Validation Accuracy: 0.9699, Loss: 0.0338
Epoch   9 Batch  116/269 - Train Accuracy: 0.9761, Validation Accuracy: 0.9641, Loss: 0.0297
Epoch   9 Batch  118/269 - Train Accuracy: 0.9676, Validation Accuracy: 0.9647, Loss: 0.0305
Epoch   9 Batch  120/269 - Train Accuracy: 0.9597, Validation Accuracy

### Save Parameters
Save the `batch_size` and `save_path` parameters for inference.

In [54]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
# Save parameters for checkpoint
helper.save_params(save_path)

# Checkpoint

In [55]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import tensorflow as tf
import numpy as np
import helper
import problem_unittests as tests

_, (source_vocab_to_int, target_vocab_to_int), (source_int_to_vocab, target_int_to_vocab) = helper.load_preprocess()
load_path = helper.load_params()

## Sentence to Sequence
To feed a sentence into the model for translation, you first need to preprocess it.  Implement the function `sentence_to_seq()` to preprocess new sentences.

- Convert the sentence to lowercase
- Convert words into ids using `vocab_to_int`
 - Convert words not in the vocabulary, to the `<UNK>` word id.

In [56]:
def sentence_to_seq(sentence, vocab_to_int):
    """
    Convert a sentence to a sequence of ids
    :param sentence: String
    :param vocab_to_int: Dictionary to go from the words to an id
    :return: List of word ids
    """
    # TODO: Implement Function
    ids = [vocab_to_int.get(word, vocab_to_int['<UNK>']) for word in sentence.lower().split()]
    return(ids)


"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_sentence_to_seq(sentence_to_seq)

Tests Passed


## Translate
This will translate `translate_sentence` from English to French.

In [57]:
translate_sentence = 'he saw a old yellow truck .'


"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
translate_sentence = sentence_to_seq(translate_sentence, source_vocab_to_int)

loaded_graph = tf.Graph()
with tf.Session(graph=loaded_graph) as sess:
    # Load saved model
    loader = tf.train.import_meta_graph(load_path + '.meta')
    loader.restore(sess, load_path)

    input_data = loaded_graph.get_tensor_by_name('input:0')
    logits = loaded_graph.get_tensor_by_name('predictions:0')
    target_sequence_length = loaded_graph.get_tensor_by_name('target_sequence_length:0')
    source_sequence_length = loaded_graph.get_tensor_by_name('source_sequence_length:0')
    keep_prob = loaded_graph.get_tensor_by_name('keep_prob:0')

    translate_logits = sess.run(logits, {input_data: [translate_sentence]*batch_size,
                                         target_sequence_length: [len(translate_sentence)*2]*batch_size,
                                         source_sequence_length: [len(translate_sentence)]*batch_size,
                                         keep_prob: 1.0})[0]

print('Input')
print('  Word Ids:      {}'.format([i for i in translate_sentence]))
print('  English Words: {}'.format([source_int_to_vocab[i] for i in translate_sentence]))

print('\nPrediction')
print('  Word Ids:      {}'.format([i for i in translate_logits]))
print('  French Words: {}'.format(" ".join([target_int_to_vocab[i] for i in translate_logits])))


INFO:tensorflow:Restoring parameters from checkpoints/dev
Input
  Word Ids:      [145, 182, 197, 200, 61, 164, 123]
  English Words: ['he', 'saw', 'a', 'old', 'yellow', 'truck', '.']

Prediction
  Word Ids:      [339, 177, 234, 304, 120, 332, 14, 231, 1]
  French Words: il a vu un vieux camion jaune . <EOS>


## Imperfect Translation
You might notice that some sentences translate better than others.  Since the dataset you're using only has a vocabulary of 227 English words of the thousands that you use, you're only going to see good results using these words.  For this project, you don't need a perfect translation. However, if you want to create a better translation model, you'll need better data.

You can train on the [WMT10 French-English corpus](http://www.statmt.org/wmt10/training-giga-fren.tar).  This dataset has more vocabulary and richer in topics discussed.  However, this will take you days to train, so make sure you've a GPU and the neural network is performing well on dataset we provided.  Just make sure you play with the WMT10 corpus after you've submitted this project.
## Submitting This Project
When submitting this project, make sure to run all the cells before saving the notebook. Save the notebook file as "dlnd_language_translation.ipynb" and save it as a HTML file under "File" -> "Download as". Include the "helper.py" and "problem_unittests.py" files in your submission.